SimFastTiming/FastTimingCommon/src/ETLDeviceSim.cc

```diff
@@ -56,22 +74,54 @@ void ETLDeviceSim::getHitsResponse(const std::vector<std::tuple<int, uint32_t, f
 
     const float toa = std::get<2>(hitRefs[i]) + tofDelay_;
     const PSimHit& hit = hits->at(hitidx);
-    const float charge = convertGeVToMeV(hit.energyLoss()) * MIPPerMeV_;
+    float charge = convertGeVToMeV(hit.energyLoss()) * MIPPerMeV_;
+
+    momentum[0] = hit.pabs();
+
+    // particle type
+    int particleType = abs(hit.particleType());
+    float MPV_ = 0;
+    if (particleType == 11) {
+      MPV_ = MPVElectron_.evaluate(momentum, emptyV);
+    } else if (particleType == 13) {
+      MPV_ = MPVMuon_.evaluate(momentum, emptyV);
+    } else if (particleType == 211) {
+      MPV_ = MPVPion_.evaluate(momentum, emptyV);
+    } else if (particleType == 321) {
+      MPV_ = MPVKaon_.evaluate(momentum, emptyV);
+    } else {
+      MPV_ = MPVProton_.evaluate(momentum, emptyV);
+    }
+    float MPV_charge = convertGeVToMeV(MPV_) * MIPPerMeV_;
 
     // calculate the simhit row and column
     const auto& position = hit.localPosition();
+
     // ETL is already in module-local coordinates so just scale to cm from mm
     Local3DPoint simscaled(convertMmToCm(position.x()), convertMmToCm(position.y()), convertMmToCm(position.z()));
+    const auto& global_point = thedet->toGlobal(simscaled);
+
+    radius[0] = global_point.perp();
+    fluence[0] = integratedLum_ * fluence_.evaluate(radius, emptyV);
+    gain[0] = lgadGain_.evaluate(fluence, emptyV);
+
     //The following lines check whether the pixel point is actually out of the active area.
-    //If that is the case it simply ignores the point but in the future some more sophisticated function could be applied.
-    if (!topo.isInPixel(simscaled)) {
-      LogDebug("ETLDeviceSim") << "Skipping hit ouf of pixel # " << hitidx << " DetId " << etlid.rawId() << " tof "
-                               << toa;
-      continue;
```

Comment:
It seems like the removal of this `continue` when `topo.isInPixel(simscaled)` could be causing the out-of-bounds crash investigated in https://github.com/cms-sw/cmssw/issues/43942

Reply:
@makortel thanks for the investigation. ETL row-colum is assigned here, based on the https://github.com/cms-sw/cmssw/blob/master/Geometry/MTDGeometryBuilder/src/RectangularMTDTopology.cc class. On the other hand the statement was modified so as to take into account a possible signal also in the interpixel gaps, i.e. the non-fully-active part of the logical partition of the surface of the detector. BTW the wrong definition of the gap size was responsible for a reduced efficiency, and was recently fixed (.i.e. too may simhits were skipped).

The point is to my mind s whether removing that ```continue``` can produce unphysical row-column values when ```topo.pixel(simscaled)``` is called. Triggered by https://github.com/cms-sw/cmssw/issues/43942#issuecomment-1941802086 , I added an assert and printed the maximum allowed range in the MTD clusterizer https://github.com/cms-sw/cmssw/blob/master/RecoLocalFastTime/FTLClusterizer/src/MTDThresholdClusterizer.cc#L73 , and this shows 

```
Buffer size [33,17]
```

i.e. 32,16 should be an allowed value that does not trigger the assert. Anyway since the assertion fires, there is clearly something wrong, by reproducing the crash on PU jobs we should be able to finally catch it.

Reply:
> I added an assert and printed the maximum allowed range in the MTD clusterizer https://github.com/cms-sw/cmssw/blob/master/RecoLocalFastTime/FTLClusterizer/src/MTDThresholdClusterizer.cc#L73 , and this shows
> 
> ```
> Buffer size [33,17]
> ```
> 
> i.e. 32,16 should be an allowed value that does not trigger the assert.

The printout
https://github.com/cms-sw/cmssw/blob/db9aceb6f549d8e8352d81ae5ec1e511af6f38bc/RecoLocalFastTime/FTLClusterizer/src/MTDThresholdClusterizer.cc#L73
adds 1 to both the number of rows and columns. The `MTDArrayBuffer::setSize()`, however, uses the number of rows and columns directly for the space
https://github.com/cms-sw/cmssw/blob/db9aceb6f549d8e8352d81ae5ec1e511af6f38bc/RecoLocalFastTime/FTLClusterizer/interface/MTDArrayBuffer.h#L134-L143
and otherwise seems to use the usual 0-based indexing
https://github.com/cms-sw/cmssw/blob/db9aceb6f549d8e8352d81ae5ec1e511af6f38bc/RecoLocalFastTime/FTLClusterizer/interface/MTDArrayBuffer.h#L170-L179
https://github.com/cms-sw/cmssw/blob/db9aceb6f549d8e8352d81ae5ec1e511af6f38bc/RecoLocalFastTime/FTLClusterizer/interface/MTDArrayBuffer.h#L108
so I'd be tempted to conclude the buffer has size for 32 x 16 elements, i.e. ranges of `[0, 31] x [0, 15]`.

Reply:
@makortel you are correct and I was wrong. My bad, I should have cross checked earlier with the geometry tests I have setup, in particular https://github.com/cms-sw/cmssw/blob/master/Geometry/MTDGeometryBuilder/test/MTDDigiGeometryAnalyzer.cc
The output for scenario D098 is quite clear:

```

 Subdetector 19 MTD Det EModule_Timingactive
 Rows     32 Columns 16 ROCS X   2 ROCS Y  1 Rows/ROC 16 Cols/ROC 16 Pitch X 0.13125 Pitch Y 0.1325 Sensor Interpad X 0.005 Sensor Interpad Y 0.005 Sensor Border X 0.005 Sensor Border Y 0.005
Pixel center location:
0 0  (-2.03437,-0.99375,0) 
0 1  (-2.03437,-0.86125,0) 
0 2  (-2.03437,-0.72875,0) 
0 3  (-2.03437,-0.59625,0) 
0 4  (-2.03437,-0.46375,0) 
0 5  (-2.03437,-0.33125,0) 
0 6  (-2.03437,-0.19875,0) 
0 7  (-2.03437,-0.06625,0) 
0 8  (-2.03437,0.06625,0) 
0 9  (-2.03437,0.19875,0) 
0 10  (-2.03437,0.33125,0) 
0 11  (-2.03437,0.46375,0) 
0 12  (-2.03437,0.59625,0) 
0 13  (-2.03437,0.72875,0) 
0 14  (-2.03437,0.86125,0) 
0 15  (-2.03437,0.99375,0) 
1 0  (-1.90312,-0.99375,0) 
1 1  (-1.90312,-0.86125,0) 
1 2  (-1.90312,-0.72875,0) 
1 3  (-1.90312,-0.59625,0) 
1 4  (-1.90312,-0.46375,0) 
1 5  (-1.90312,-0.33125,0) 
1 6  (-1.90312,-0.19875,0) 
1 7  (-1.90312,-0.06625,0) 
1 8  (-1.90312,0.06625,0) 
1 9  (-1.90312,0.19875,0) 
1 10  (-1.90312,0.33125,0) 
1 11  (-1.90312,0.46375,0) 
1 12  (-1.90312,0.59625,0) 
1 13  (-1.90312,0.72875,0) 
1 14  (-1.90312,0.86125,0) 
1 15  (-1.90312,0.99375,0) 
2 0  (-1.77187,-0.99375,0) 
2 1  (-1.77187,-0.86125,0) 
(...)
30 14  (1.90312,0.86125,0) 
30 15  (1.90312,0.99375,0) 
31 0  (2.03437,-0.99375,0) 
31 1  (2.03437,-0.86125,0) 
31 2  (2.03437,-0.72875,0) 
31 3  (2.03437,-0.59625,0) 
31 4  (2.03437,-0.46375,0) 
31 5  (2.03437,-0.33125,0) 
31 6  (2.03437,-0.19875,0) 
31 7  (2.03437,-0.06625,0) 
31 8  (2.03437,0.06625,0) 
31 9  (2.03437,0.19875,0) 
31 10  (2.03437,0.33125,0) 
31 11  (2.03437,0.46375,0) 
31 12  (2.03437,0.59625,0) 
31 13  (2.03437,0.72875,0) 
31 14  (2.03437,0.86125,0) 
31 15  (2.03437,0.99375,0) 

```

as I was initially suggesting in https://github.com/cms-sw/cmssw/issues/43942#issuecomment-1943801653 . The clusterizer buffer is larger than the allowed real range, this will have to be possibly reviewed, and in any case adding an assert check is useful to catch immediately possible issues. Even in the ```RectangularMTDTopology``` in first place. 
I'll check how to preserve the treatment of the inter-pixel active area gap that @parbol has added without spoiling the index of the pixel itself.